### PR TITLE
Improve performance of Unit division

### PR DESCRIPTION
**Pull request by exupero**

### CODE REVIEW

File: src/main/java/javax/measure/unit/ProductUnit.java
Comment (oehme) [inline src/main/java/javax/measure/unit/ProductUnit.java:191]: We should check equals, not identity.

**Review by oehme — COMMENTED** [review]
Do you have some benchmark results?

File: src/main/java/javax/measure/unit/ProductUnit.java
Comment (afbarbaro) [reply src/main/java/javax/measure/unit/ProductUnit.java:191]: Checking identity here should be good enough, right? since we now cache instances of `Unit` subclasses.

**Comment by oehme** [comment]
> There seem to be much bigger gains from just using doGetInstance within getQuotientInstance:

We always need to go through the cache when creating new units, so we get the same unit inside a SignalSegment, rather than a new unit instance for each Sample. 

The problem with this benchmark is that it only shows you the cost of the initial division, but never uses that unit again for any follow-up computations. So you get a skewed perception of it. That said, I think we can make this faster:

I [profiled](https://github.com/seeq12/jscience/files/9434256/divideby.html.zip) this scenario and the bottleneck is `Unit.inverse`, which I'm sure we can improve. One solution would be to remember the inverse of each unit in a field, just like we did for the `standardUnit`.

**Comment by oehme** [comment]
Also, I'm not sure why we're hitting the `inverse` code path here at all. In the SAME_UNITS case we should have `leftUom==rightUom`, but that's not happening. Might be a problem in the benchmark setup.

**Comment by oehme** [comment]
Okay, I understand the problem with the benchmark now. We first divide e.g. `m/m`. That gives us `Unit.ONE`. Then we divide that by `m` again, which results in the `unit.inverse` call, with the result `1/m`.

File: src/main/java/javax/measure/unit/ProductUnit.java
Comment (oehme) [reply src/main/java/javax/measure/unit/ProductUnit.java:191]: We only cache ProductUnit, none of the others. But as mentioned further in the PR, this doesn't solve the issue and there's a better solution.

**Comment by oehme** [comment]
I've pushed up changes that make this a lot faster. Please have a look @exupero

**Comment by exupero** [comment]
Nice finds, that made a big improvement:
benchmarks

**Review by oehme — APPROVED** [review]
Thanks for checking, this is good to go from my side then.

**Comment by exupero** [comment]
@afbarbaro any comments before I merge?

**Comment by afbarbaro** [comment]
> @afbarbaro any comments before I merge?

Sorry I didn't get to this before - I reviewed and it LGTM!

**Pull request merged**
